src/validators/literal.rs

```diff
@@ -69,32 +75,20 @@ impl<T: Debug> LiteralLookup<T> {
                     .as_cow()
                     .map_err(|_| py_schema_error_type!("error extracting str {:?}", k))?;
                 expected_str.insert(str.to_string(), id);
+                expected_py_primitives.push((k.as_unbound().clone_ref(py), id));
             } else if expected_py_dict.set_item(&k, id).is_err() {
                 expected_py_values.push((k.as_unbound().clone_ref(py), id));
             }
         }
 
         Ok(Self {
-            expected_bool: match expected_bool.true_id.is_some() || expected_bool.false_id.is_some() {
-                true => Some(expected_bool),
-                false => None,
-            },
-            expected_int: match expected_int.is_empty() {
-                true => None,
-                false => Some(expected_int),
-            },
-            expected_str: match expected_str.is_empty() {
-                true => None,
-                false => Some(expected_str),
-            },
-            expected_py_dict: match expected_py_dict.is_empty() {
-                true => None,
-                false => Some(expected_py_dict.into()),
-            },
-            expected_py_values: match expected_py_values.is_empty() {
-                true => None,
-                false => Some(expected_py_values),
-            },
+            expected_bool: (expected_bool.true_id.is_some() || expected_bool.false_id.is_some())
+                .then_some(expected_bool),
+            expected_int: (!expected_int.is_empty()).then_some(expected_int),
+            expected_str: (!expected_str.is_empty()).then_some(expected_str),
+            expected_py_dict: (!expected_py_dict.is_empty()).then_some(expected_py_dict.into()),
+            expected_py_values: (!expected_py_values.is_empty()).then_some(expected_py_values),
+            expected_py_primitives: (!expected_py_primitives.is_empty()).then_some(expected_py_primitives),
```

Comment:
I wonder, is there an advantage of `expected_py_primitives` over `expected_py_dict`?

Reply:
Conclusion was that `expected_py_primitives` will be used only in `lax` mode. I think it _probably_ makes sense to store as a dict and use hash lookup (as that checks `__hash__` _and_ `__eq__`) - we can always make more lax later.

Reply:
Also conclusion - we're not going to fall back to union validation, that's too lax + time consuming.

Reply:
Alright - can't enforce strict / lax here as we're in the literal lookup, but I've added the hash check.

Reply:
If enforcing the hash check, surely a dict is just straight up more performant?